design/ceph/admission-controller.md

```diff
@@ -0,0 +1,182 @@
+
+## Certificate Management
+
+### Development and Testing 
+
+![Dev_AC_Image](./ac_dev_design.png)
+
+We can use self-signed certificates approved by the Kubernetes Certificate Authority for development purposes.This can be done by following the steps given below.
+* Creating the private key and certs using openssl.
+
+**Sample : Generating the public and private keys** 
+
+```
+openssl genrsa -out ${PRIVATE_KEY_NAME}.pem 2048
+ openssl req -new -key ${PRIVATE_KEY_NAME}.pem -subj "/CN=${service}.${namespace}.svc"
+
+#Created after CSR is approved
+openssl base64 -d -A -out ${PUBLIC_KEY_NAME}.pem 
+```
+
+* Creating and sending a CSR to kubernetes for approval
+
+**Sample : Certificate Signing Request (CSR) in Kubernetes**
+```
+apiVersion: certificates.k8s.io/v1beta1
+kind: CertificateSigningRequest
+metadata:
+  name: ${csrName}
+spec:
+  request: $(cat server.csr | base64 | tr -d '\n')
+  usages:
+  - digital signature
+  - key encipherment
+  - server auth
+```
+Verify with
+` kubectl get csr ${csrName}`
+
+**Sample : Approval of Signed Certificate**
+
+```
+kubectl certificate approve ${csrName}
+```
+
+If it is approved, then we can check the certificate with following command 
+
+```
+$(kubectl get csr ${csrName} -o jsonpath='{.status.certificate}')
+```
+
+* Once approved, a generic secret will be created with the given public and private key which will be later mounted onto the server pod for use.
+
+**Sample : Creating a Secret in Kubernetes**
+```
+kubectl create secret generic ${secret} \
+        --from-file=key.pem=${PRIVATE_KEY_NAME}.pem \
+        --from-file=cert.pem=${PUBLIC_KEY_NAME}.pem
+```
+* Modifying the webhook config to inject CA bundle onto the ValidatingWebhookConfig
+* All the above resources will be created in rook-ceph namespace
+
+
+Using the above approach, the dev/admins will have the responsibility of rotating the certificates when they expire.
+
+Below is an excerpt of what a ValidatingWebhookConfig looks like 
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: demo-webhook
+webhooks:
+  - name: webhook-server.webhook-demo.svc
+    clientConfig:
+      service:
+        name: webhook-server
+        namespace: rook-ceph
+        path: "/validate"
+      caBundle: ${CA_PEM_B64}
+    rules:
+      - operations: [ "CREATE" ]
+        apiGroups: [""]
+        apiVersions: ["v1"]
+        resources: ["pods"]
+```
+
+We can make changes to the above values according to intercept based on whether a resource is being updated/deleted/created or change the type of resource or the request path which will be sent to the server.
+
+## Setup in Rook
+
+### Starting the Webhook Server
+
+The webhook server binaries will be bundled with rook operator image. Based on the whether the relevant secrets are present, the rook operator will automatically trigger the webhook binaries, deploy the relevant configuration files onto the cluster and start the server. 
```

Comment:
When you have a prototype of this I'm curious how much the image size will increase with the webhook binaries. I would have hoped Rook would just be referencing libraries for starting the https server and there wouldn't be new binaries to pull into the image.

Reply:
There would definitely be an increase an image size we may want to avoid.  I have modified my statement here. The rook operator itself would be starting an HTTP server (through libraries) when the appropriate secrets are detected.  Thanks for your suggestion